### ISSUE TITLE
Train on custom dataset

### ISSUE THREAD

**Issue by VyBui** [NathanUA/U-2-Net#76]
Hello @NathanUA,

Great work on the model!

I can download the code and run prediction and the results are AMAZING. 

Now I want to train the model on the custom dataset. Basically, I have an image with 11 classes. Here is what I have modified your code:

1) Before feeding the mask/label image to the model. I have split the image into 11 items:
`masks = ([(label == v) for v in range(11)])
mask = np.concatenate(masks, axis=2).astype('float')`
in line 258 right before data_loader.py

2) `net = U2NET(3, 1)`` to `net = U2NET(3, 11)` 

3) `self.outconv = nn.Conv2d(6,out_ch,1)` to `self.outconv = nn.Conv2d(66,out_ch,1)`

More information: the label images have 11 values which are from zero (0) to ten (10). Each value represents the segmentation of each part of the object.

The training loss and tar seem correct (see the attached image) 
![Screenshot from 2020-10-05 16-18-47](https://user-images.githubusercontent.com/15938335/95062068-a565dc80-0726-11eb-884d-eb15166b1c93.png)

However, the results are not what I expected. Something is not right 
![2_0_Jeans](https://user-images.githubusercontent.com/15938335/95062231-ea8a0e80-0726-11eb-95c5-3ce574c9fe22.png)
![2_1_Jeans](https://user-images.githubusercontent.com/15938335/95062238-ebbb3b80-0726-11eb-825b-a161e2371e19.png)
![2_10_Jeans](https://user-images.githubusercontent.com/15938335/95062242-ec53d200-0726-11eb-9b03-eb72b01cfae2.png)
![3_7_Jeans](https://user-images.githubusercontent.com/15938335/95062245-ecec6880-0726-11eb-8ec0-004a8aa133f1.png)
![4_0_Jeans](https://user-images.githubusercontent.com/15938335/95062251-ed84ff00-0726-11eb-9d93-ff8283dfb17e.png)
![4_1_Jeans](https://user-images.githubusercontent.com/15938335/95062252-ee1d9580-0726-11eb-9850-da3e6719e661.png)
![4_2_Jeans](https://user-images.githubusercontent.com/15938335/95062255-ee1d9580-0726-11eb-86cd-b71d10aef8b7.png)
![5_2_Jeans](https://user-images.githubusercontent.com/15938335/95062259-eeb62c00-0726-11eb-8e33-7af0ea179d9b.png)
![6_0_Jeans](https://user-images.githubusercontent.com/15938335/95062260-ef4ec280-0726-11eb-98ea-7f86903619b4.png)
![6_2_Jeans](https://user-images.githubusercontent.com/15938335/95062262-efe75900-0726-11eb-8276-3cc7b8e26862.png)
![7_1_Jeans](https://user-images.githubusercontent.com/15938335/95062264-efe75900-0726-11eb-9e80-da3ce7cee396.png)
![7_2_Jeans](https://user-images.githubusercontent.com/15938335/95062267-f07fef80-0726-11eb-840f-f8edcab35232.png)

I would like to ask if you have already faced this similar issue before, on top of your head is there anything I must have done incorrectly? 

Looking forward to hearing from you.

Vy

**Comment by PytaichukBohdan** [comment]
@VyBui hello!

Maybe this would help :) The zeros on the image usually represent background, when 1 (255 as white pixels) - foreground. 
Have you tried switching the labels? 

Btw, which custom dataset are you using?
I'm currently working on a very similar problem, we can cooperate :)

**Comment by VyBui** [comment]
Hey @PytaichukBohdan, 
Thank you for your comment, The labels contain 11 classes which range between from 0 to 10 in the 0-255(uint8) setting. 
I believe it has nothing to do with the label, it is because I have configured something wrong.

Thank you for your question, we have a custom dataset for a single human (include hands, legs, tops, bottoms, accessories ..) in an image. Which is similar to the ATR dataset. 

If you have any question or want to cooperate, drop an email to me at buithevybkdn@gmail.com
Thanks

**Comment by i-amgeek** [comment]
Hey @VyBui 

As you can see in u2net_train.py, by default model is using BinaryCrossEntropy Loss. It expects binary target values. i.e. 0 and 1. 

I think you should use nn.CrossEntropyLoss instead.

**Comment by VyBui** [comment]
Hey @i-amgeek,

Thank you for your solution, I am finally able to train the network with multiples classes now. I will update the loss function accordingly to your advice.

I will update you soon.

Best regards

**Comment by lmarson94** [comment]
Hi, 
I am facing similar issues to yours: I trained the model from scratch to segment people from images on a custom dataset (~5000 images), but with only one class so that the masks are either 0 or 1.

Do you have any insights on why sometimes the predicted masks look similar to the one posted by the OP?

**Comment by PytaichukBohdan** [comment]
@VyBui can you please elaborate on how did you adapt the code for nn.CrossEntropyLoss?

**Comment by i-amgeek** [comment]
Hey @VyBui, changing loss function worked?
